src/NUnitTestAdapter/TestingPlatformAdapter/NUnitBridgedTestFramework.cs

```diff
@@ -0,0 +1,115 @@
+            : base(extension, getTestAssemblies, serviceProvider, capabilities)
+        {
+            PatchNUnit3InternalLoggerBug();
+        }
+
+        /// <inheritdoc />
+        protected override Task SynchronizedDiscoverTestsAsync(VSTestDiscoverTestExecutionRequest request, IMessageBus messageBus,
+            CancellationToken cancellationToken)
+        {
+            new NUnit3TestDiscoverer()
+                .DiscoverTests(request.AssemblyPaths, request.DiscoveryContext, request.MessageLogger, request.DiscoverySink);
+
+            return Task.CompletedTask;
+        }
+
+        /// <inheritdoc />
+        protected override Task SynchronizedRunTestsAsync(VSTestRunTestExecutionRequest request, IMessageBus messageBus,
+            CancellationToken cancellationToken)
+        {
+            ITestExecutor executor = new NUnit3TestExecutor();
+            using (cancellationToken.Register(executor.Cancel))
+            {
+                if (request.VSTestFilter.TestCases is { } testCases)
+                {
+                    executor.RunTests(testCases, request.RunContext, request.FrameworkHandle);
+                }
+                else
+                {
+                    executor.RunTests(request.AssemblyPaths, request.RunContext, request.FrameworkHandle);
+                }
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private static void PatchNUnit3InternalLoggerBug()
+        {
+            // NUnit3 will internally call InternalTracer.Initialize(...), and it will check that it is not called multiple times.
+            // We call it multiple times in server mode, in indeterminate order (it does not always have to be Discover and then Run),
+            // When InternalTracer.Initialize is called with Off, it will not set traceWriter internal field to anything, but when you call it again
+            // it will try to write "Initialize was already called" via this writer without null check and will fail with NRE.
+            //
+            // Patch for this was implemented in NUnit4, but won't be backported to NUnit3 because we are hitting a bit of an edge case:
+            // https://github.com/nunit/nunit/issues/4255
+            //
+            // To fix us, we set the writer to a null writer, so any subsequent calls to Initialize will write to null instead of failing.
+            // We also need to do this under a lock, and not rely on the InternalTracer.Initialized, because that might be set to late and we would
+            // re-enter the method twice.
+            if (initialized)
+            {
+                return;
+            }
+
+            lock (InitializationLock)
+            {
+                if (initialized)
+                {
+                    return;
+                }
+
+                // TODO: Uncomment this line when InternalsVisibleTo is set up.
+                // var nopWriter = new InternalTraceWriter(new StreamWriter(Stream.Null));
```

Comment:
I am not sure what's different but on our last test with NUnit we were able to access this type.

Reply:
Yes, it is public. And has been so since 2013.

Reply:
@OsirisTerje This is not public, the class is internal
![image](https://github.com/nunit/nunit3-vs-adapter/assets/11340282/ed202d95-0456-43f5-8228-31cdac3ab337)

We need you to set the `InternalsVisibleTo` from nunit.engine.core to testadapter.

Reply:
My bad, I just checked the old code and we were actually referencing `nunit` package in addition to `nunit.engine`.

Reply:
We're talking about two different InternalTraceWriter classes, I referred to the one in the framework:
![image](https://github.com/nunit/nunit3-vs-adapter/assets/203432/9745fd4b-fa1f-45e4-a5a6-8ae5c0754025)

Referencing the framework from the adapter might be troublesome.  We need to ensure people can use different versions  of nunit with the same adapter.  Possibly some clever use of wildcards on the versions can handle it. 

Changing the engine to make this public instead of internal is doable.

Reply:
I honestly don't remember well all details but we used this trick following a bug we were having. @nohwnd Do you recall?

Reply:
I don't remember, but all the details of the fix, including link to the issue it is fixing are in the code in form of comments. 🙂 If we need to change our dependencies and use class that is not public, we instantiate it via reflection as well, since we are already patching it via reflection. But looks like you reverted the `// new ...` so I guess it is okay as is.